### PR TITLE
Added helper loadData to facilitate Unique Service Selectors validation between manifest objects

### PR DESCRIPTION
**Pull request by tbarnes94**
# Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Adding this helper and new for loop preprocesses the manifest objects for any violations concerning duplicate naming of selectors.

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Is it a  breaking change which will impact consuming tool(s).

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pkg/safeguards/tests/unique-service-selectors/USS-error-manifest.yaml [hunk 1]
```diff
@@ -1,5 +1,27 @@
 apiVersion: apps/v1
-kind: Deployment
+kind: Service
+metadata:
+  name: my-deployment
```
[the remark this answers is not in the document]
Comment (tbarnes94) [reply pkg/safeguards/tests/unique-service-selectors/USS-error-manifest.yaml:4]: fixed

File: pkg/safeguards/tests/unique-service-selectors/USS-success-manifest.yaml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 apiVersion: apps/v1
-kind: Deployment
+kind: Service
 metadata:
   name: my-deployment
```
[the remark this answers is not in the document]
Comment (tbarnes94) [reply pkg/safeguards/tests/unique-service-selectors/USS-success-manifest.yaml:4]: fixed

**Review by NickKeller — APPROVED** [review]

File: pkg/safeguards/tests/unique-service-selectors/USS-error-manifest.yaml [hunk 3]
```diff
@@ -1,7 +1,29 @@
...
+kind: Service
 metadata:
-  name: my-deployment
+  name: my-deployment-1
+  labels:
+    app: my-app
+    environment: production
+    testLabel3: randomlabel
+spec:
+  replicas: 3
```
Comment (NickKeller) [inline pkg/safeguards/tests/unique-service-selectors/USS-error-manifest.yaml:10]: services don't have replicas

File: pkg/safeguards/tests/unique-service-selectors/USS-success-manifest.yaml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 apiVersion: apps/v1
-kind: Deployment
+kind: Service
 metadata:
-  name: my-deployment
+  name: my-deployment-1
```
Comment (NickKeller) [inline pkg/safeguards/tests/unique-service-selectors/USS-success-manifest.yaml:4]: take out all deployment related fields

**Review by NickKeller — CHANGES_REQUESTED** [review]

File: cmd/validate.go [hunk 5]
```diff
@@ -91,14 +95,17 @@ func (vc *validateCmd) run(c *cobra.Command) error {
 		return fmt.Errorf("not a valid file or directory: %w", err)
 	}
 
-	var manifestFiles []string
+	var manifestFiles []safeguards.ManifestFile
 	if isDir {
```
Comment (davidgamero) [inline cmd/validate.go:99]: is this isDir branch necessary? the description of walkdir and walkdirtree say it should run on a single file as well as a directory recursively
> walks the file tree rooted at root, calling fn for each file or directory in the tree, **including root** [link](https://pkg.go.dev/path/filepath#Walk)
this means we can likely just always call the code in the isDir case

File: cmd/validate.go [hunk 6]
```diff
@@ -119,18 +126,20 @@ func (vc *validateCmd) run(c *cobra.Command) error {
...
 		}
 		if !manifestViolationsFound {
-			log.Printf("    no violations found.")
+			log.Printf("    ✅ no violations found.")
 		}
 	}
 
 	if len(manifestViolations) > 0 {
 		c.SilenceUsage = true
-		return fmt.Errorf("violations found in manifests")
```
Comment (davidgamero) [inline cmd/validate.go:133]: we can delete this line if the output is going to be removed

File: pkg/safeguards/safeguards_constants.go [hunk 7]
```diff
@@ -27,6 +27,16 @@ type Safeguard struct {
 	constraintPath string
 }
 
+type ManifestFile struct {
```
Comment (davidgamero) [inline pkg/safeguards/safeguards_constants.go:30]: are these constants? they seem like types so the file name could be misleading

File: pkg/safeguards/safeguards.go [hunk 8]
```diff
@@ -68,23 +64,40 @@ func GetManifestViolations(ctx context.Context, manifestFiles []string) ([]Manif
...
+	// organized map of manifest object by file name
+	manifestMap := make(map[string][]*unstructured.Unstructured, 0)
+
+	// aggregate of every manifest object into one list
+	objects := []*unstructured.Unstructured{}
+
 	for _, m := range manifestFiles {
-		var objectViolations map[string][]string
-		objects, err := fc.ReadManifests(m) // read all the objects stored in a single file
+		objs, err := fc.ReadManifests(m.Path) // read all the objects stored in a single file
```
Comment (davidgamero) [inline pkg/safeguards/safeguards.go:74]: nit: having objs and objects in the same scope could benefit from more precise variable names

File: pkg/safeguards/safeguards.go [hunk 9]
```diff
@@ -68,23 +64,40 @@ func GetManifestViolations(ctx context.Context, manifestFiles []string) ([]Manif
...
 			return manifestViolations, err
 		}
 
+		objects = append(objects, objs...)
+		manifestMap[m.Name] = objs
+	}
+
+	// thbarnes: loadData loads manifest data into client for review as well
+	if len(objects) > 0 {
+		err = loadData(ctx, c, objects)
```
Comment (davidgamero) [inline pkg/safeguards/safeguards.go:86]: nit: renaming this method to something more precise would negate the need for a comment explaining where data is being loaded to/from

**Review by davidgamero — CHANGES_REQUESTED** [review]

File: pkg/safeguards/safeguards_constants.go (hunk 7 above)
Comment (tbarnes94) [reply pkg/safeguards/safeguards_constants.go:30]: you're right

File: pkg/safeguards/tests/unique-service-selectors/USS-error-manifest.yaml [hunk 10]
```diff
@@ -1,21 +1,23 @@
...
+    app.kubernetes.io/name: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+---
+apiVersion: apps/v1
+kind: Service
+metadata:
+  name: my-service-2
```
Comment (NickKeller) [inline pkg/safeguards/tests/unique-service-selectors/USS-error-manifest.yaml:16]: you should also explicitly specify the namespace on all the service objects

File: pkg/safeguards/safeguards.go (hunk 8 above)
Comment (tbarnes94) [reply pkg/safeguards/safeguards.go:74]: fixed

File: pkg/safeguards/safeguards.go (hunk 9 above)
Comment (tbarnes94) [reply pkg/safeguards/safeguards.go:86]: fixed

File: cmd/validate.go (hunk 5 above)
Comment (tbarnes94) [reply cmd/validate.go:99]: I will look into this in a future PR

File: pkg/safeguards/tests/unique-service-selectors/USS-error-manifest.yaml (hunk 10 above)
Comment (tbarnes94) [reply pkg/safeguards/tests/unique-service-selectors/USS-error-manifest.yaml:17]: fixed

**Review by NickKeller — APPROVED** [review]

File: pkg/safeguards/safeguards_test.go [hunk 11]
```diff
@@ -183,30 +183,30 @@ func TestValidateDeployment_ContainerLimits(t *testing.T) {
...
-//	err = loadConstraints(ctx, c, []*unstructured.Unstructured{constraint})
-//	assert.Nil(t, err)
-//
-//	// validating deployment manifests
-//	validateTestManifests_Fail(ctx, t, c, testFc, testManifest_USS.ErrorPaths)
-//	validateTestManifests_Success(ctx, t, c, testFc, testManifest_USS.SuccessPaths)
-//}
-//
+
+// TODO: rich description here
```
Comment (davidgamero) [inline pkg/safeguards/safeguards_test.go:187]: is this TODO meant to still be here?

**Pull request merged**